pcdsdevices/mirror.py

```diff
@@ -385,10 +400,216 @@ class XOffsetMirror(BaseInterface, GroupDevice):
+        """
+        Calculate the lightpath-state information by dispatching to
+        helper methods.  These methods may change in later subclasses,
+        as may the types of the arguments.
+
+        However, the base logic will remain the same.
+        (the way we walk the decision tree does not change)
+
+        We assume that in every case, we receive the following info
+        - x-position --> insertion status
+        - y-position --> coating information
+        - pitch      --> output branch
+
+        Roughly speaking the decision tree goes as follows:
+        - Determine if mirror is inserted:
+            - if removed, beam goes straight through, return
+            - if inserted, continue
+        - Determine which coating is being used, and use coating to
+          determine which set of pitch ranges to use
+        - Determine which destination the mirror is delivering beam to
+          based on the pitch
+        """
+        try:
+            x_out, x_in = self._get_insertion_state(x_info)
+
+            if x_out:
+                return LightpathState(inserted=x_in, removed=x_out,
+                                      output={self.output_branches[0]: 1})
+
+            # if in, check coating
+            coating_index = self._get_coating_index(y_info)
+
+            out_branch = self._get_output_branch(coating_index, pitch_info)
+
+            # transmission is always 1 if miror is configured properly
+            trans = 1
+
+            return LightpathState(
+                inserted=x_in,
+                removed=x_out,
+                output={out_branch: trans}
+            )
+        except MirrorLogicError as ex:
+            # a state for if calculation cannot proceed
+            self.log.debug(ex)
+            return LightpathState(
+                inserted=False,
+                removed=False,
+                output={self.output_branches[0]: 0}
+            )
+
+    def _get_insertion_state(self, x: float) -> Tuple[bool, bool]:
+        """
+        Interpret x-position as inserted or removed, based on ranges
+        provided at init.
+
+        Assumes first range is OUT, second range is IN
+
+        Parameters
+        ----------
+        x : float
+            x-position
+
+        Raises
+        ------
+        MirrorLogicError
+            if the provided insertion ranges are malformed (wrong shape)
+
+        Returns
+        -------
+        is_out, is_in : Tuple[bool, bool]
+            tuple of booleans describing the inserted and removed status
+        """
+        if self.x_ranges == []:
+            # default case for always-in mirrors
+            return False, True
+
+        if np.shape(self.x_ranges) != (2, 2):
+            # improper ranges for insertion, fail
+            raise MirrorLogicError(
+                'Provided x-ranges are the malformed. '
+                f'got: {np.shape(self.x_ranges)}, expected (2,2)')
+
+        ins_bools = [(limit[0] < x) and (x < limit[1])
```

Comment:
Python syntactic sugar instead, perhaps?
```suggestion
        ins_bools = [limit[0] < x < limit[1]
```